quinn-proto/src/range_set.rs

```diff
@@ -138,6 +138,33 @@ impl RangeSet {
         before || after
     }
 
+    /// Add a range to the set, returning the ranges it replaced
+    pub fn insert_drain(&mut self, mut range: Range<u64>) -> InsertDrain<'_> {
+        let pred = if let Some((prev_start, prev_end)) = self.pred(range.start) {
```

Comment:
Is there some way to prevent the 3-level cascading `else { None }` here? (Something with `Option::filter()`, I suppose.)

(Also, a little documentation on `succ()` and, in particular `pred()`, might help. I guess I haven't looked at this code in much detail before, but `pred` primarily invoked "predicate" for me rather than `predecessor`.)

Don't love this name. Would it make sense to just call this `insert()`? It looks like there's some duplication between `insert()` and `insert_drain()` currently, so if there's not too much overhead it might make sense to not maintain both.

Reply:
> Is there some way to prevent the 3-level cascading else { None } here?

We could do this with combinators, but the main block of code here has side effects on the data structure, and I'm leery of hiding that. Still, was able to unnest by one level, which is something.

> documentation

:+1: 

> Don't love this name.

I agree the name isn't great. How's `replace`?

> Would it make sense to just call this insert()?

I'm reluctant because, at least syntatically, we go to a bunch of extra effort to produce an iterator. This did prompt me to reexamine and simplify `insert`, though, which makes keeping it that much more pleasant.

Reply:
Much better!